ProjFS.Mac/PrjFSKextTests/MockProc.cpp

```diff
@@ -2,22 +2,50 @@
 #include <string>
 
 static std::string s_procName;
+static int s_processPid = 1;
+static int s_kauth_cred_getuid_count = 0;
+static proc_t s_proc_find_result = nullptr;
+static bool s_servicePidOnFirstCall = false;
 
-int proc_pid(proc_t)
+void ResetMockProc()
 {
-    return 1;
+    s_processPid = 1;
+    s_proc_find_result = nullptr;
+    s_kauth_cred_getuid_count = 0;
+    s_servicePidOnFirstCall = false;
 }
 
-void proc_name(int pid, char* buf, int size)
+void ReturnServicePidOnFirstCall()
```

Comment:
This was the easiest way to get a service pic on the first (and only the first) call.

Reply:
I had some trouble following how the mock process code works, as well as understanding exactly what these helpers do and when to call them:

```
void ResetMockProc();
void SetProcName(const std::string& procName);
void SetPpid(int pid);
void SetDummyProcFind();
void ReturnServicePidOnFirstCall();
```

What would you think if rather than controlling the results of each mock function we built up a collection of mock processes/credentials (similar to what we're doing for mock vnodes)?

The interface could then be something like:

```
MockProc_AddCredential(int credentialId, uid_t UID);
MockProc_AddProcess(int pid, int credentialId, int ppid);
MockProc_SetSelfPid(int pid); // Optional, we could also have a well known SELF_PID value
MockProc_Reset();
```

And the test would look like:

```
    // Fail the test for a service pid (processUID >= 500), and get a non valid Parent Pid ( pid <= 1)
    MockProc_AddCredential(1 /*credentialId*/, 1 /*UID*/);
    MockProc_AddProcess(5 /*pid*/, 1 /*credentialId*/, 1 /*ppid*/);
    MockProc_SetSelfPid(5);
    XCTAssertFalse(CurrentProcessWasSpawnedByRegularUser());
    MockProc_Reset();

...

   // Fail the test for a service pid (processUID >= 500), and get a valid Parent  Pid / Process with a non-service Pid
    MockProc_AddCredential(1 /*credentialId*/, 1 /*UID*/);
    MockProc_AddCredential(2 /*credentialId*/, 501 /*UID*/);
    MockProc_AddProcess(5 /*pid*/, 1 /*credentialId*/, 6 /*ppid*/); 
    MockProc_AddProcess(6 /*pid*/, 2 /*credentialId*/, 0 /*ppid*/);
    MockProc_SetSelfPid(5);
    XCTAssertTrue(CurrentProcessWasSpawnedByRegularUser());
    MockProc_Reset();
```

Which would also allow for larger process hierarchies:

```
    MockProc_AddCredential(1 /*credentialId*/, 1 /*UID*/);
    MockProc_AddCredential(2 /*credentialId*/, 501 /*UID*/);

    // Force CurrentProcessWasSpawnedByRegularUser to walk several levels
    MockProc_AddProcess(5 /*pid*/, 1 /*credentialId*/, 6 /*ppid*/); 
    MockProc_AddProcess(6 /*pid*/, 1 /*credentialId*/, 7 /*ppid*/);
    MockProc_AddProcess(7 /*pid*/, 1 /*credentialId*/, 8 /*ppid*/);
    MockProc_AddProcess(8 /*pid*/, 2 /*credentialId*/, 0 /*ppid*/);
    MockProc_SetSelfPid(5);
    XCTAssertTrue(CurrentProcessWasSpawnedByRegularUser());
    MockProc_Reset();
```

Internally we could just keep collections of the credentials and processes:

```
struct proc
{
    int pid;
    int credentialId;
    int ppid;
}

static map<int /*credential ID*/, int /* UID */> credentialMap;
static map<int /*process ID*/, proc> processMap;
static int selfPid;
```

Reply:
`MockProc_AddProcess(int pid, int credentialId, int ppid);` could have another parameter for process name as well

Reply:
I agree that's clearer I'll update accordingly.

Reply:
@wilbaker I've updated per your suggestion please let me know what you think!